Libraries/LibWeb/Clipboard/ClipboardItem.cpp

```diff
@@ -0,0 +1,209 @@
+                    // 1. If v is a DOMString, then follow the below steps:
+                    if (value.is_string()) {
+                        // 1. Let dataAsBytes be the result of UTF-8 encoding v.
+                        auto utf8_string = value.as_string().utf8_string();
+                        auto data_as_bytes = MUST(ByteBuffer::copy(utf8_string.bytes()));
+
+                        // 2. Let blobData be a Blob created using dataAsBytes with its type set to mimeType, serialized.
+                        auto blob_data = FileAPI::Blob::create(realm, data_as_bytes, mime_type_serialized);
+
+                        // 3. Resolve p with blobData.
+                        WebIDL::resolve_promise(realm, promise, blob_data);
+                    }
+                    // 2. If v is a Blob, then follow the below steps:
+                    if (value.is_object() && is<FileAPI::Blob>(value.as_object())) {
+                        // 1. Resolve p with v.
+                        WebIDL::resolve_promise(realm, promise, value);
+                    }
+
+                    return JS::js_undefined();
+                }),
+                // 2. If representationDataPromise was rejected, then:
+                GC::create_function(realm.heap(), [&realm, type, promise](JS::Value) -> WebIDL::ExceptionOr<JS::Value> {
+                    // 1. Reject p with "NotFoundError" DOMException in realm.
+                    WebIDL::reject_promise(realm, promise, WebIDL::NotFoundError::create(realm, MUST(String::formatted("No data found for MIME type: {}", type))));
+
+                    return JS::js_undefined();
+                }));
+
+            // 3. Return p.
+            return promise;
+        }
+    }
+
+    // 9. Reject p with "NotFoundError" DOMException in realm.
+    WebIDL::reject_promise(realm, promise, WebIDL::NotFoundError::create(realm, MUST(String::formatted("No data found for MIME type: {}", type))));
+
+    // 10. Return p.
+    return promise;
+}
+
+// https://w3c.github.io/clipboard-apis/#dom-clipboarditem-supports
+bool ClipboardItem::supports(JS::VM&, String const& type)
+{
+    // 1. If type is in mandatory data types or optional data types, then return true.
+    // 2. If not, then return false.
+    // TODO: Implement optional data types, like web custom formats and image/svg+xml.
+    return any_of(MANDATORY_DATA_TYPES, [&](auto supported) { return supported == type; });
+}
+
+ClipboardItem::ClipboardItem(JS::Realm& realm)
+    : Bindings::PlatformObject(realm)
+{
+}
+
+ClipboardItem::~ClipboardItem() = default;
+
+void ClipboardItem::initialize(JS::Realm& realm)
+{
+    Base::initialize(realm);
+    WEB_SET_PROTOTYPE_FOR_INTERFACE(ClipboardItem);
+}
+
+void ClipboardItem::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    for (auto& representation : m_representations) {
+        visitor.visit(representation.data);
+    }
```

Comment:
Visiting `representation.data` isn't necessary given that its type is `GC::Root<WebIDL::Promise>`. `GC::Root` objects are self-visiting.

Reply:
Indeed, and we should really avoid storing `GC::Root` in member variables like this,  since it's extremely easy for that to cause GC leaks. (The promise ("representation") will keep the realm alive, and if the promise is reachable from anywhere in the realm, we now leak the entire realm.)

Better to store a `Vector<GC::Ref<WebIDL::Promise>>` and visit the promises, which avoids this whole problem :)

Reply:
> Better to store a Vector<GC::Ref<WebIDL::Promise>> and visit the promises, which avoids this whole problem :)

@awesomekling I've changed `data` from `GC::Root<WebIDL::Promise>` to `GC::Ref<WebIDL::Promise>`. Can this change avoid GC leaks problem? or I need to use exactly `Vector<GC::Ref<WebIDL::Promise>>`? 

```C++
    struct Representation {
        String mime_type;              // The MIME type (e.g., "text/plain").
        bool is_custom;                // Whether this is a web custom format.
        GC::Ref<WebIDL::Promise> data; // The actual data for this representation.
    };
    
private:
    Vector<Representation> m_representations;
```